Source/santabs/SNTBundleService.m

```diff
@@ -0,0 +1,275 @@
+  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
+    [self createConnection];
+  });
+}
+
+- (void)hashBundleBinariesForEvent:(SNTStoredEvent *)event
+                             reply:(SNTBundleHashBlock)reply {
+  NSProgress *progress =
+      [NSProgress currentProgress] ? [NSProgress progressWithTotalUnitCount:1] : nil;
+
+  NSDate *startTime = [NSDate date];
+
+  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
+    NSArray *relatedBinaries = [self findRelatedBinaries:event progress:progress];
+    NSString *bundleHash = [self calculateBundleHashFromEvents:relatedBinaries];
+    NSNumber *ms = [NSNumber numberWithDouble:[startTime timeIntervalSinceNow] * -1000.0];
+    if (bundleHash) LOGD(@"hashed %@ in %@ ms", event.fileBundlePath, ms);
+    reply(bundleHash, relatedBinaries, ms);
+  });
+}
+
+#pragma mark Internal Methods
+
+/**
+ Find binaries within a bundle given the bundle's event. Will run until completion, however long
+ that might be, or until the NSProgress is cancelled. Search is done within the bundle concurrently.
+
+ @param event The SNTStoredEvent to begin searching underneath
+ @return An array of SNTStoredEvent's
+ */
+- (NSArray *)findRelatedBinaries:(SNTStoredEvent *)event progress:(NSProgress *)progress {
+  // For storing the generated events, with a simple lock for writing.
+  NSMutableArray *relatedEvents = [NSMutableArray array];
+
+  // For storing files to be hashed
+  NSMutableSet<SNTFileInfo *> *fis = [NSMutableSet set];
+
+  // Limit the number of threads that can process files at once to keep CPU usage down.
+  dispatch_semaphore_t sema =
+      dispatch_semaphore_create([[NSProcessInfo processInfo] processorCount] / 2);
+
+  // Group the processing into a single group so we can wait on the whole group at the end.
+  dispatch_group_t group = dispatch_group_create();
+
+  // Directory enumerator
+  NSDirectoryEnumerator *dirEnum =
+      [[NSFileManager defaultManager] enumeratorAtPath:event.fileBundlePath];
+
+  // Locks for accessing the enumerator and adding file and events between threads.
+  __block pthread_mutex_t enumeratorMutex = PTHREAD_MUTEX_INITIALIZER;
+  __block pthread_mutex_t eventsMutex = PTHREAD_MUTEX_INITIALIZER;
+
+  // Counts used as additional progress information in Santa GUI
+  __block uint64_t binaryCount = 0;
+  __block uint64_t sentBinaryCount = 0;
+  __block uint64_t fileCount = 0;
+
+  __block BOOL breakDir = NO;
+
+  // In the first stage iterate over every file in the tree checking if it is a binary. If so add
+  // it to the fis set for the second stage. Hashing the file while iterating over the filesystem
+  // causes some weird performance issues. Do them separately.
+  while (1) {
+    @autoreleasepool {
+      if (breakDir || progress.isCancelled) break;
+
+      // Wait for a processing thread to become available
+      dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
```

Comment:
Waiting on processing threads forever should be fine. These are non-blocking from the rest of the bundle logic. We will have a timeout for the entire bundle hashing.